### ISSUE TITLE
YOLOv5 loss

### ISSUE THREAD

**Issue by MZeinab** [ultralytics/yolov5#11712]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

@glenn-jocher
Hi, could you please write the exact formula that you used in loss.py, I don't understand what exactly it is,
it will be so useful for us and even for adding it to our articles. I checked the loss.py, if I understood correctly you mean this as loss:

` def call(self, p, targets): # predictions, targets
lcls = torch.zeros(1, device=self.device) # class loss
lbox = torch.zeros(1, device=self.device) # box loss
lobj = torch.zeros(1, device=self.device) # object loss
tcls, tbox, indices, anchors = self.build_targets(p, targets) # targets

    # Losses
    for i, pi in enumerate(p):  # layer index, layer predictions
        b, a, gj, gi = indices[i]  # image, anchor, gridy, gridx
        tobj = torch.zeros(pi.shape[:4], dtype=pi.dtype, device=self.device)  # target obj

        n = b.shape[0]  # number of targets
        if n:
            # pxy, pwh, _, pcls = pi[b, a, gj, gi].tensor_split((2, 4, 5), dim=1)  # faster, requires torch 1.8.0
            pxy, pwh, _, pcls = pi[b, a, gj, gi].split((2, 2, 1, self.nc), 1)  # target-subset of predictions

            # Regression
            pxy = pxy.sigmoid() * 2 - 0.5
            pwh = (pwh.sigmoid() * 2) ** 2 * anchors[i]
            pbox = torch.cat((pxy, pwh), 1)  # predicted box
            iou = bbox_iou(pbox, tbox[i], CIoU=True).squeeze()  # iou(prediction, target)
            lbox += (1.0 - iou).mean()  # iou loss

            # Objectness
            iou = iou.detach().clamp(0).type(tobj.dtype)
            if self.sort_obj_iou:
                j = iou.argsort()
                b, a, gj, gi, iou = b[j], a[j], gj[j], gi[j], iou[j]
            if self.gr < 1:
                iou = (1.0 - self.gr) + self.gr * iou
            tobj[b, a, gj, gi] = iou  # iou ratio

            # Classification
            if self.nc > 1:  # cls loss (only if multiple classes)
                t = torch.full_like(pcls, self.cn, device=self.device)  # targets
                t[range(n), tcls[i]] = self.cp
                lcls += self.BCEcls(pcls, t)  # BCE

            # Append targets to text file
            # with open('targets.txt', 'a') as file:
            #     [file.write('%11.5g ' * 4 % tuple(x) + '\n') for x in torch.cat((txy[i], twh[i]), 1)]

        obji = self.BCEobj(pi[..., 4], tobj)
        lobj += obji * self.balance[i]  # obj loss
        if self.autobalance:
            self.balance[i] = self.balance[i] * 0.9999 + 0.0001 / obji.detach().item()

    if self.autobalance:
        self.balance = [x / self.balance[self.ssi] for x in self.balance]
    lbox *= self.hyp['box']
    lobj *= self.hyp['obj']
    lcls *= self.hyp['cls']
    bs = tobj.shape[0]  # batch size

    return (lbox + lobj + lcls) * bs, torch.cat((lbox, lobj, lcls)).detach()`

I trained my custom model with your friendly tutorials, yolov5 v6.0, I want to write a report about it and explain the exact loss function, but I am really confused. I'll appreciate it if you could guide.
Best of all,

**Comment by glenn-jocher** [comment]
@MZeinab hi there 👋,

The loss function you've pasted is from YOLOv5 loss.py, which calculates the combined box, object, and class loss. If you'd like to understand the detailed mathematical formulation, I'd recommend referring to the YOLOv5 documentation at https://docs.ultralytics.com/yolov5. 

The loss function encompasses several components such as objectness, classification, and regression, and accounts for factors like class balance and optimization hyperparameters. It's important to understand and communicate these concepts accurately in your report.

If you have any specific questions or need further assistance, feel free to ask. Happy to help!

Best regards,

**Comment by LapTQ** [comment]
@glenn-jocher Hi there, I would like to ask about the implementation details of the objectness loss. But first, would you mind confirm if I understand correctly:

1. Given a list of GT boxes, the `build_targets` method will assign each GT box to every anchor-boxes on the grid that fits (one GT box can be assigned to many anchor-boxes).
2. Given these assigned anchor-boxes, the `__call__` method will get the prediction result corresponding to them, then compute IoU with the GT boxes from which the anchor-boxes were assigned. This IoU score is used as the objectness score stored in the `tobj` variable.
3. For the other anchor-boxes that was NOT assigned, the objectness scores are set as `0` in `tobj`.
4. The objectness loss is computed as BCE loss between the predicted objectness and the `tobj`.

If I understand all those correctly, then I wonder whether, in step 2, choosing the IoU as the objectness score to store in `tobj` is adequate? Why don't we store 1 in `tobj` for these assigned anchor-boxes?

Thank you!
